### PR TITLE
GH-37378: [C++] Add A Dictionary Compaction Function For DictionaryArray

### LINKED ISSUE #37378: [C++] Add a Dictionary Compaction Function For DictionaryArray
### Describe the enhancement requested

DictionaryArray could have unused dictionary values. This is an obstacle to finish #37100. A Dictionary Compaction Function, which removes unused values in dictionary, is useful to solve the problem. Besides, #37277 has the same request in Go implementation.

Here is a pesudo example:
```
dict_a
dictionary: ["1", "2", "3"]
indices: [0, 1, 0, 1]

dictionary_compaction(dict_a)
dictionary: ["1", "2"]
indices: [0, 1, 0, 1]
```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by R-JunmingChen**
### Rationale for this change
A Dictionary Compaction Function for DictionaryArray is supported.

### What changes are included in this PR?
Add a Function for Dictionary Compaction

### Are these changes tested?
Yes

Are there any user-facing changes?
No
* Closes: #37378

### CODE REVIEW

File: cpp/src/arrow/compute/registry.cc [hunk 1]
```diff
@@ -276,6 +276,7 @@ static std::unique_ptr<FunctionRegistry> CreateBuiltInRegistry() {
   // Register core kernels
   RegisterScalarCast(registry.get());
   RegisterDictionaryDecode(registry.get());
+  RegisterDictionaryCompaction(registry.get());
```
[the remark this answers is not in the document]
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/registry.cc]: Sorry, I have no consideration on these.

> should that be done in the #if ARROW_COMPUTE block below?

Have updated

> Granted, I'm pretty sure doing that would also necessitate moving the tests out of vector_hash_test.cc since it's built/run unconditionally.

Sure, actully, I have no idea where to place the codes. Do we need to create a new file `vector_dictionary.cc` and `vector_dictionary_test.cc`?

[the remark this answers is not in the document]
Comment (benibus) [reply cpp/src/arrow/compute/registry.cc]: > Sorry, I have no consideration on these.

Understandable. The whole situation isn't really ideal at the moment...

> Do we need to create new files `vector_dictionary.cc` and `vector_dictionary_test.cc`?

You _could_ get away with only moving the tests, but creating `vector_dictionary.cc` would be better. Just make sure it's added to the appropriate section in the CMakeLists: https://github.com/R-JunmingChen/arrow/blob/46959c21d2ca6a2bdd8026e407c90a34b907fdb6/cpp/src/arrow/CMakeLists.txt#L434

[the remark this answers is not in the document]
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/registry.cc]: I add `vector_dictionary.cc` and `vector_dictionary_test.cc`.

File: cpp/src/arrow/compute/kernels/vector_dictionary_test.cc [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                                       const ArrayVector& expected) {
+  auto input_chunked_array = std::make_shared<ChunkedArray>(input, dict_type);
+
+  ASSERT_OK_AND_ASSIGN(Datum actual_datum, DictionaryCompaction(input_chunked_array));
+  ValidateOutput(actual_datum);
+  auto actual = actual_datum.chunked_array();
+  AssertChunkedEqual(*actual, expected);
+}
+
+TEST_F(TestDictionaryCompactionKernel, DictionaryArray) {
```
Comment (benibus) [inline cpp/src/arrow/compute/kernels/vector_dictionary_test.cc:61]: We should probably add a test case where the index array has a nonzero offset - e.g. it was sliced from a larger array.

File: cpp/src/arrow/compute/kernels/vector_dictionary_test.cc [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    CheckDictionaryCompaction(dict_type, "[null]", "[]", "[]", "[]");
+    CheckDictionaryCompaction(dict_type, "[false]", "[null]", "[]", "[null]");
+    CheckDictionaryCompaction(dict_type, "[true, false]", "[0]", "[true]", "[0]");
+    CheckDictionaryCompaction(dict_type, "[true, null, false]", "[2, 1]", "[null, false]",
+                              "[1, 0]");
+    CheckDictionaryCompaction(dict_type, "[true, false]", "[0, null]", "[true]",
+                              "[0, null]");
+    CheckDictionaryCompaction(dict_type, "[true, null, false]", "[2, null, 1]",
+                              "[null, false]", "[1, null, 0]");
+  }
```
Comment (benibus) [inline cpp/src/arrow/compute/kernels/vector_dictionary_test.cc:90]: It'd be nice to also add a test case with a larger input array, where several contiguous chunks of the dictionary array are unused. It'd stress the logic for index re-adjustment more.

**Review by benibus — COMMENTED** [review]
Thanks! The implementation looks pretty good to me. Just a few comments.

File: cpp/src/arrow/compute/api_vector.cc [hunk 4]
```diff
@@ -327,6 +327,10 @@ Result<Datum> DictionaryEncode(const Datum& value, const DictionaryEncodeOptions
   return CallFunction("dictionary_encode", {value}, &options, ctx);
 }
 
+Result<Datum> DictionaryCompaction(const Datum& value, ExecContext* ctx) {
+  return CallFunction("dictionary_compaction", {value}, ctx);
```
[the remark this answers is not in the document]
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_vector.cc:331]: indeed, let me update

File: cpp/src/arrow/compute/kernels/vector_dictionary_test.cc (hunk 2 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary_test.cc]: have added

File: cpp/src/arrow/compute/kernels/vector_dictionary_test.cc (hunk 3 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary_test.cc:166]: Sure, I add some tests and a comment for that.

File: cpp/src/arrow/compute/kernels/vector_dictionary_test.cc (hunk 2 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary_test.cc]: Have updated

**Review by benibus — APPROVED** [review]

**Comment by R-JunmingChen** [comment]
Hi, @benibus, do you know any reviewer can merge this PR, so we can go on the review and check in process.

File: cpp/src/arrow/compute/api_vector.h [hunk 5]
```diff
@@ -586,6 +586,20 @@ Result<Datum> DictionaryEncode(
     const DictionaryEncodeOptions& options = DictionaryEncodeOptions::Defaults(),
     ExecContext* ctx = NULLPTR);
 
+/// \brief Compact a dictionary array
+///
+/// The output removes unused values in dictionary from the input.
+/// The function assumes every indice is effective.
```
Comment (bkietz) [inline cpp/src/arrow/compute/api_vector.h:592]: Could you reword this to be more explicit about what is assumed about each index? I'm not sure what you mean by "effective"

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+namespace {
+
+// Dictionary compaction implementation
+
+const FunctionDoc dictionary_compact_doc{
+    "Compact dictionary array",
+    ("Return a compacted version of the dictionary array input,\n"
+     "which removes unused values in dictionary.\n"
+     "The function assumes every indice is effective."),
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:39]: Here too

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+ public:
+  Result<std::shared_ptr<Array>> Exec(std::shared_ptr<Array> dict_array,
+                                      ExecContext* ctx) const override {
+    const DictionaryArray& casted_dict_array =
+        checked_cast<const DictionaryArray&>(*dict_array);
+    const std::shared_ptr<Array>& dict = casted_dict_array.dictionary();
+    if (dict->length() == 0) {
+      return dict_array;
+    }
+    const std::shared_ptr<Array>& indice = casted_dict_array.indices();
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:62]:
```suggestion
    const std::shared_ptr<Array>& indices = casted_dict_array.indices();
```

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+    // check whether the input is compacted
+    std::vector<bool> dict_used(dict->length(), false);
+    int64_t dict_used_count = 0;
+    for (int64_t i = 0; i < indice->length(); i++) {
+      if (indice->IsNull(i)) {
+        continue;
+      }
+
+      CType cur_indice = indices_data[i + offset];
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:80]:
```suggestion
      CType current_index = indices_data[i + offset];
```

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    CType len = static_cast<CType>(dict->length());
+    for (CType i = 0; i < len; i++) {
+      if (dict_used[i]) {
+        dict_indice.push_back(i);
+      } else if (i + 1 < len && dict_used[i + 1]) {
+        need_change_indice = true;
+      }
+    }
+    BuilderType dict_indice_builder;
+    ARROW_RETURN_NOT_OK(dict_indice_builder.AppendValues(dict_indice));
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:108]: Wouldn't it be more straightforward to append to the builder in the loop instead of building a vector first?

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::Array> compacted_dict_indices,
+                          dict_indice_builder.Finish());
+    ARROW_ASSIGN_OR_RAISE(
+        auto compacted_dict_res,
+        Take(dict, compacted_dict_indices, TakeOptions::NoBoundsCheck(), ctx));
+    std::shared_ptr<arrow::Array> compacted_dict = compacted_dict_res.make_array();
+
+    // indice changes
+    if (!need_change_indice) {
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:117]: Please check this case before building the vector of indices; that way we might be able to skip the allocation

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    }
+    const std::shared_ptr<Array>& indice = casted_dict_array.indices();
+    if (indice->length() == 0) {
+      ARROW_ASSIGN_OR_RAISE(auto empty_dict,
+                            MakeEmptyArray(dict->type(), ctx->memory_pool()));
+      return DictionaryArray::FromArrays(dict_array->type(), indice, empty_dict);
+    }
+    const CType* indices_data =
+        reinterpret_cast<const CType*>(indice->data()->buffers[1]->data());
+    int64_t offset = indice->data()->offset;
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:70]: The GetValues helper function can reduce the boilerplate here. It also adds in the offset
```suggestion
    const auto* indices_data = indices->data()->GetValues<CType>(1);
```

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+      indice_minus_number[0] = 1;
+    }
+    for (int64_t i = 1; i < dict->length(); i++) {
+      indice_minus_number[i] = indice_minus_number[i - 1];
+      if (!dict_used[i]) {
+        indice_minus_number[i] = indice_minus_number[i] + 1;
+      }
+    }
+
+    std::vector<CType> raw_changed_indice(indice->length(), 0);
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:131]: here too please use the builder directly rather than building vectors

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+    std::vector<bool> is_valid(indice->length(), true);
+    for (int64_t i = 0; i < indice->length(); i++) {
+      if (indice->IsNull(i)) {
+        is_valid[i] = false;
+      } else {
+        CType cur_indice = indices_data[i + offset];
+        raw_changed_indice[i] = cur_indice - indice_minus_number[cur_indice];
+      }
+    }
+    BuilderType indice_builder;
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:141]: Please use the context's memory pool; it may have overridden the default
```suggestion
    BuilderType indices_builder(ctx->memory_pool());
```

**Review by bkietz — CHANGES_REQUESTED** [review]
Thanks for working on this!

File: cpp/src/arrow/compute/api_vector.h (hunk 5 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_vector.h:592]: Sure, I change to the following words:
The function assumes every indice has its corresponding value  in dictionary.

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc (hunk 9 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary.cc:107]: Sure, let me change the code

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc (hunk 10 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary.cc:116]: Hi, it seems that this check has been already before the process of building the vector of indices if you mean the `indices_builder` below.

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    ARROW_ASSIGN_OR_RAISE(
+        auto compacted_dict_res,
+        Take(dict, compacted_dict_indices, TakeOptions::NoBoundsCheck(), ctx));
+    std::shared_ptr<arrow::Array> compacted_dict = compacted_dict_res.make_array();
+
+    // indices changes
+    if (!need_change_indice) {
+      return DictionaryArray::FromArrays(dict_array->type(), indices, compacted_dict);
+    }
+    std::vector<CType> indice_minus_number(dict->length(), 0);
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:117]: I think you can accomplish this more simply with another call to `Take`: just build an array mapping `compacted_dict_indices[i] -> i`

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+
+namespace {
+
+// Dictionary compaction implementation
+
+const FunctionDoc dictionary_compact_doc{
+    "Compact dictionary array",
+    ("Return a compacted version of the dictionary array input,\n"
+     "which removes unused values in dictionary.\n"
+     "The function assumes every indices has its corresponding value\n"
```
[the remark this answers is not in the document]
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary.cc:39]: make senses, I add a bound check.

File: cpp/src/arrow/compute/api_vector.h [hunk 16]
```diff
@@ -586,6 +586,19 @@ Result<Datum> DictionaryEncode(
...
 
+/// \brief Compact a dictionary array
+///
+/// The output removes unused values in dictionary from the input.
+///
+/// \param[in] value array-like input, which should be a dictionary type.
+/// \param[in] ctx the function execution context, optional
+/// \return compacted dictionary array
+///
+/// \since 13.0.0
```
Comment (mapleFU) [inline cpp/src/arrow/compute/api_vector.h:597]: Should we change it to since 14.0.0?

**Review by mapleFU — COMMENTED** [review]
So this function is separated into 3 steps:
1. Extract a bitmap for used dictionary from dict indices
2. Build a `compact_dict` using `Take`
3. Remap original `dict_indices` to new `changed_indices`?

The 1-2 looks good to me, 3 is a bit confusing. Actually `indice_minus_number` is a map for `dict_indices`?

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+     "in dictionary."),
+    {"dictionary_array"}};
+
+class DictionaryCompactKernel : public KernelState {
+ public:
+  virtual Result<std::shared_ptr<Array>> Exec(std::shared_ptr<Array> dict_array,
+                                              ExecContext* ctx) const = 0;
+};
```
[the remark this answers is not in the document]
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary.cc:45]: Hi, I change to `Exec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out)`. Does it ok? Since there is error when I use `ArrayKernelExec exec =  Exec` in DictionaryCompactKernelImpl<T>.

One curious question, for the old version code, the data is warpped by `std::shared_ptr<Array>`. Since it's a pointer, the performance could be Ok? Is there some subtle places to influence the performance?

**Comment by R-JunmingChen** [comment]
> 3. Remap original `dict_indices` to new `changed_indices`?

remap the `indices` to new `changed_indices`. You can check the test for better understanding

```
    // input isn't compacted && its indice needs to be adjusted
    CheckDictionaryCompact(dict_type, "[3, 4, 7, 0, 12, 191, 21, 8]",
                           "[0, 2, 4, 4, 6, 4, 2, 0, 6]", "[3, 7, 12, 21]",
                           "[0, 1, 2, 2, 3, 2, 1, 0, 3]");
    CheckDictionaryCompact(dict_type, "[3, 4, 7, 0, 12, 191, 21, 8]",
                           "[4, 6, 7, 7, 6, 4, 6, 6, 6]", "[12, 21, 8]",
                           "[0, 1, 2, 2, 1, 0, 1, 1, 1]");
    CheckDictionaryCompact(dict_type, "[3, 4, 7, 0, 12, 191, 21, 8]",
                           "[7, 4, 7, 7, 7, 7, 4, 7, 7]", "[12, 8]",
                           "[1, 0, 1, 1, 1, 1, 0, 1, 1]");
```

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+      return DictionaryArray::FromArrays(dict_array->type(), indices, empty_dict);
+    }
+    const CType* indices_data = indices->data()->GetValues<CType>(1);
+
+    // check whether the input is compacted
+    std::vector<bool> dict_used(dict->length(), false);
+    int64_t dict_used_count = 0;
+    for (int64_t i = 0; i < indices->length(); i++) {
+      if (indices->IsNull(i)) {
+        continue;
```
[the remark this answers is not in the document]
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary.cc:83]: Hi, another curious question, why the helpers is more efficient than a for loop? since the lambda function have some extra heads in runtime.

Besides, with a for loop, we can end in advance when `dict_used_count == dict->length()`.
```
        if (dict_used_count == dict->length()) {  // input is already compacted
          out->value = dict_array->data();
          return Status::OK();
        }
```

I think it's ok for the following code to make this change:
```
    for (int64_t i = 0; i < indices->length(); i++) {
      if (indices->IsNull(i)) {
        ARROW_RETURN_NOT_OK(indices_builder.AppendNull());
      } else {
        CType current_index = indices_data[i];
        ARROW_RETURN_NOT_OK(
            indices_builder.Append(current_index - index_minus_number[current_index]));
      }
    }
```

**Comment by mapleFU** [comment]
Oh I see this mapping accumulates unused dict item and give the origin count that a number it should minus, got that.

File: cpp/src/arrow/compute/api_vector.h (hunk 16 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/api_vector.h:597]: Sure

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc (hunk 14 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/compute/kernels/vector_dictionary.cc]: Sorry, I don't get your idea. could you please describe it with more details?

**Comment by pitrou** [comment]
This doesn't need to be a compute function. There are already dict-related utilities in `src/arrow/array/array_dict.h` (including `DictionaryArray::Transpose` which is certainly useful here).

So I would recommend something like:
```c++
class DictionaryArray {
  /// Return a transpose map to remove any unused dictionary entries.
  Result<std::vector<int32_t>> Compact() const;
};
```

Or perhaps (similar idea but using the Arrow C++ allocators rather than the STL):
```c++
class DictionaryArray {
  /// Return a transpose map to remove any unused dictionary entries.
  /// \result the transpose map as a buffer of int32 values
  Result<std::shared_ptr<Buffer>> Compact() const;
};
```

**Comment by R-JunmingChen** [comment]
Hi, @pitrou, I don't get the point. 
From my understanding, `the DictionaryArray::Transpose` can be used to map original indices to new indices, and the new dictionary should be taken as input. Do you mean we compact dictionary by a way firstly and then use `DictionaryArray::Transpose` to map the indices?

**Comment by bkietz** [comment]
> This doesn't need to be a compute function.

I'm sorry I hadn't proposed this possibility already; it's certainly much less complex if it's tenable. Since this issue was opened as a prerequisite of https://github.com/apache/arrow/pull/37100 I was unnecessarily stuck on on handling this problem in the context of compute kernels.

I think providing `DictionaryArray::Compact` will be acceptable. To flesh out the proposal a little more:

```c++
  /// Return an equivalent dictionary array with no unused dictionary entries.
  ///
  /// For example, consider a dictionary array where only two values of the dictionary
  /// are referenced:
  ///    dict_array = {indices=[3, 5, 3, 5], dictionary=["a", "b", "c", "d", "e", "f"]}
  ///    assert dict_array.Compact() == {indices=[0, 1, 0, 1], dictionary=["d", "f"]}
  Result<std::shared_ptr<DictionaryArray>> DictionaryArray::Compact(MemoryPool*) const;

  /// Return transpose mapping from uncompacted indices to compacted indices and a
  ///
  /// For example, consider a dictionary array where only two values of the dictionary
  /// are referenced:
  ///    dict_array = {indices=[3, 5, 3, 5], dictionary=["a", "b", "c", "d", "e", "f"]}
  ///    map = dict_array.CompactTransposeIndices()
  ///    assert map == [-1, -1, -1, 0, -1, 1]
  ///    assert dict_array.Transpose(map) == dict_array.Compact()
  Result<std::shared_ptr<Buffer>> DictionaryArray::CompactTransposeIndices(MemoryPool*) const;
```

**Comment by R-JunmingChen** [comment]
Ok, I get the point, thank guidances from @bkietz and @pitrou.

Before I implement this version, I have aquestion to discuss:
Do we really need to make `DictionaryArray::CompactTransposeIndices` a member function? `DictionaryArray::Compact` seems to be adequate to be exposed to users, or we make `DictionaryArray::CompactTransposeIndices` a private member function?

**Comment by bkietz** [comment]
I think making it a private member function will be acceptable. It will be easy enough to promote it to public if there's a need

**Comment by R-JunmingChen** [comment]
Ok, let's change to this version

**Comment by R-JunmingChen** [comment]
Hi @pitrou @bkietz, I have implemented a draft based on what we discuessed. Codes is in `array_dict_test.cc`, `array_dict.h` and `array_dict.cc`. I will add more tests once the implementation is agreed by you.

**Comment by R-JunmingChen** [comment]
Besides, after implementing this version, here are some disvantages I think need to be discussed:

1. Performance
The running of the Function version can be ended in advance. To be specific, the dictionary array which is compacted can be directly returned after checking. However, the Member Function version still need to construct a transpose map and call  `Transpose`.

2.Dependency
Both version we need to call `Take`. But I am not sure whether it's good for `dict_array.cc` to include `arrow/compute/api.h`, since the compilation of `compute` is optional.

File: cpp/src/arrow/array/array_dict.cc [hunk 19]
```diff
@@ -211,6 +212,92 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
...
+  std::vector<bool> dict_used(dict_length, false);
+  CType dict_len = static_cast<CType>(dict_length);
+  for (int64_t i = 0; i < index_length; i++) {
+    if (data->IsNull(i)) {
+      continue;
+    }
+
+    CType current_index = indices_data[i];
+    if (current_index < 0 || current_index >= dict_len) {
+      return Status::IndexError("indice out of bound:", current_index);
```
Comment (bkietz) [inline cpp/src/arrow/array/array_dict.cc:238]:
```suggestion
      return Status::IndexError("Index out of bounds while compacting dictionary array: ", current_index, "(dictionary is ", dict_length, " long) at position ", i);
```

File: cpp/src/arrow/array/array_dict.cc [hunk 20]
```diff
@@ -211,6 +212,92 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
...
+    } else if (!dict_used[current_index]) {
+      dict_used[current_index] = true;
+    }
```
Comment (bkietz) [inline cpp/src/arrow/array/array_dict.cc:241]: This `else` is unnecessary
```suggestion
    }

    dict_used[current_index] = true;
```

File: cpp/src/arrow/array/array_dict.cc [hunk 21]
```diff
@@ -211,6 +212,92 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
...
+  }
+
```
Comment (bkietz) [inline cpp/src/arrow/array/array_dict.cc:243]: I think it'd be acceptable for this function to return a null transpose map to indicate that transposition is unnecessary. In that case you can recover the fast path for cases where the dictionary is already compacted.
```suggestion
  CType max_used_index = 0, used_count = 0;
  for (auto [i, used] : Zip(Enumerate<CType>, dict_used)) {
    if (used) {
      max_used_index = i;
      ++used_count;
    }
  }
  if (max_used_index == used_count - 1) {
    // The dictionary is already compact, so just return here
    out_compact_dictionary = MakeArray(data);
    return nullptr;
  }

```

File: cpp/src/arrow/array/array_dict.cc [hunk 22]
```diff
@@ -211,6 +212,92 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
...
+Result<std::unique_ptr<Buffer>> CompactTransposeMap(
+    const std::shared_ptr<ArrayData>& data, MemoryPool* pool,
+    std::shared_ptr<Array>& out_compact_dictionary) {
+  if (data->type->id() != Type::DICTIONARY) {
+    return Status::TypeError("Expected dictionary type");
+  }
+
+  const auto& dict_type = checked_cast<const DictionaryType&>(*data->type);
+  switch (dict_type.index_type()->id()) {
```
Comment (bkietz) [inline cpp/src/arrow/array/array_dict.cc:279]: Please rewrite this to use VisitTypeInline instead of a switch

**Review by bkietz — CHANGES_REQUESTED** [review]
This looks great, thanks for persevering! I have a few comments on the new function.

File: cpp/src/arrow/array/array_dict.cc (hunk 22 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/array/array_dict.cc]: updated

File: cpp/src/arrow/array/array_dict.cc (hunk 21 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/array/array_dict.cc:289]: I reuse the old version code to check compaction in advance

File: cpp/src/arrow/array/array_dict.cc [hunk 23]
```diff
@@ -211,6 +212,105 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
...
+      if (data->IsNull(i)) {
+        continue;
+      }
+
+      CType current_index = indices_data[i];
+      if (current_index < 0 || current_index >= dict_len) {
+        return Status::IndexError(
+            "Index out of bounds while compacting dictionary array: ", current_index,
+            "(dictionary is ", dict_length, " long) at position ", i);
+      } else if (!dict_used[current_index]) {
```
Comment (bkietz) [inline cpp/src/arrow/array/array_dict.cc:251]:
```suggestion
      }
      if (dict_used[current_index]) continue;
```

File: cpp/src/arrow/array/array_dict.cc [hunk 24]
```diff
@@ -211,6 +212,105 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
...
+      }
+    }
+
+    using BuilderType = NumericBuilder<IndexArrowType>;
+    using arrow::compute::Take;
+    using arrow::compute::TakeOptions;
+    BuilderType dict_indices_builder(pool);
+    ARROW_ASSIGN_OR_RAISE(output_map,
+                          AllocateBuffer(dict_length * sizeof(int32_t), pool));
+    int32_t* output_map_raw = reinterpret_cast<int32_t*>(output_map->mutable_data());
```
Comment (bkietz) [inline cpp/src/arrow/array/array_dict.cc:270]:
```suggestion
    auto* output_map_raw = output_map->mutable_data_as<int32_t>();
```

File: cpp/src/arrow/array/array_dict.cc [hunk 25]
```diff
@@ -211,6 +212,105 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
...
+      CType current_index = indices_data[i];
+      if (current_index < 0 || current_index >= dict_len) {
+        return Status::IndexError(
+            "Index out of bounds while compacting dictionary array: ", current_index,
+            "(dictionary is ", dict_length, " long) at position ", i);
+      } else if (!dict_used[current_index]) {
+        dict_used[current_index] = true;
+        dict_used_count++;
+
+        if (dict_used_count == dict_length) {
```
Comment (bkietz) [inline cpp/src/arrow/array/array_dict.cc:255]: Checking here enables skipping the rest of the dictionary, which is good. However I think it'd also be useful to detect usage of only a slice of the dictionary. If you'd prefer not to handle that in this PR, please write a follow up issue

File: cpp/src/arrow/array/array_dict.cc [hunk 26]
```diff
@@ -211,6 +212,105 @@ Result<std::shared_ptr<ArrayData>> TransposeDictIndices(
...
+    using arrow::compute::Take;
+    using arrow::compute::TakeOptions;
+    BuilderType dict_indices_builder(pool);
+    ARROW_ASSIGN_OR_RAISE(output_map,
+                          AllocateBuffer(dict_length * sizeof(int32_t), pool));
+    int32_t* output_map_raw = reinterpret_cast<int32_t*>(output_map->mutable_data());
+    int32_t current_index = 0;
+    for (CType i = 0; i < dict_len; i++) {
+      if (dict_used[i]) {
+        ARROW_RETURN_NOT_OK(dict_indices_builder.Append(i));
```
Comment (bkietz) [inline cpp/src/arrow/array/array_dict.cc:274]: Since we know the number of indices ahead of time, please presize the Builder and use unsafeappend

File: cpp/src/arrow/compute/kernels/vector_dictionary.cc [hunk 27]
```diff
@@ -0,0 +1,213 @@
...
+namespace arrow {
+
+namespace compute {
+namespace internal {
+
+namespace {
+
+// Dictionary compaction implementation
+
+const FunctionDoc dictionary_compact_doc{
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_dictionary.cc:35]: The compute function should be removed

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/array/array_dict.cc (hunk 25 above)
Comment (mapleFU) [reply cpp/src/arrow/array/array_dict.cc]: Can we just use `Slice` outside the `Compact` to handling this?

File: cpp/src/arrow/array/array_dict_test.cc [hunk 28]
```diff
@@ -1428,6 +1428,71 @@ TEST(TestDictionary, IndicesArray) {
...
+TEST(TestDictionary, Compact) {
+  std::shared_ptr<arrow::DataType> type;
+  std::shared_ptr<arrow::DataType> dict_type;
+
+  for (const auto& index_type : all_dictionary_index_types()) {
+    ARROW_SCOPED_TRACE("index_type = ", index_type->ToString());
+
+    type = boolean();
+    dict_type = dictionary(index_type, type);
+
```
Comment (mapleFU) [inline cpp/src/arrow/array/array_dict_test.cc:1455]: Should we add a test for invalid input type?

**Review by mapleFU — COMMENTED** [review]
Rest LGTM, thanks!

**Review by bkietz — APPROVED** [review]
LGTM, thanks for working on this!

File: cpp/src/arrow/array/array_dict_test.cc (hunk 28 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/array/array_dict_test.cc:1455]: No. We can't create a DictionaryArray with invalid index type.

Comment (mapleFU) [reply cpp/src/arrow/array/array_dict_test.cc:1455]: Oh I see a 

```c++
  if (data->type->id() != Type::DICTIONARY) {
    return Status::TypeError("Expected dictionary type");
  }
```

Here, but seems it cannot be called

**Review by mapleFU — APPROVED** [review]

**Comment by R-JunmingChen** [comment]
Strange. All C++ tests are canceled in CI.

**Comment by bkietz** [comment]
CI failures look unrelated. Thanks!

File: cpp/src/arrow/array/array_dict.cc (hunk 25 above)
Comment (R-JunmingChen) [reply cpp/src/arrow/array/array_dict.cc]: I write a follow up issue #38247 to handle the optimization
